core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java

```diff
@@ -639,6 +607,39 @@ public MutationWriter(int numSendThreads) {
       binningThreadPool =
           ThreadPools.createFixedThreadPool(1, "BinMutations", new SynchronousQueue<>(), false);
       binningThreadPool.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+      binningThreadPool.execute(Trace.wrap(new Runnable() {
+        @Override
+        public void run() {
+          final MutationSet readyForBinning = new MutationSet();
+          while (!closed) {
+            int num = incomingQueue.size() + readyForBinning.size();
+            if (num > 0) {
+              log.trace("{} - binning {} mutations", Thread.currentThread().getName(), num);
+              try {
+                TableIdAndMutation tm = null;
+                while ((tm = incomingQueue.poll()) != null) {
+                  readyForBinning.add(tm.getTableId(), tm.getMutation());
+                }
+                addMutations(readyForBinning);
+                readyForBinning.clear();
+              } catch (Exception e) {
+                updateUnknownErrors("Error processing mutation set", e);
+              }
+            } else {
+              log.trace("{} - waiting for mutations to arrive", Thread.currentThread().getName());
+              try {
+                TableIdAndMutation tm = incomingQueue.take();
+                readyForBinning.add(tm.getTableId(), tm.getMutation());
+              } catch (InterruptedException e) {
+                if (!closed) {
+                  log.error("sleep interrupted", e);
+                }
+              }
+            }
```

Comment:
Could do the following to handle close while waiting for data and handle all exceptions in one place.  

```suggestion
            try {
              TableIdAndMutation tm;
              while ((tm = incomingQueue.poll(100, TimeUnit.MILLISECONDS)) == null) {
                if (closed)
                  return;
              }
              readyForBinning.add(tm.getTableId(), tm.getMutation());
              while ((tm = incomingQueue.poll()) != null) {
                  readyForBinning.add(tm.getTableId(), tm.getMutation());
              }
              log.trace("{} - binning {} mutations", Thread.currentThread().getName(),
								readyForBinning.size());
              addMutations(readyForBinning);
              readyForBinning.clear();
            } catch (InterruptedException e) {
               if(!closed)
                  updateUnknownErrors("Error processing mutation set", e);
            } catch (Exception e) {
                updateUnknownErrors("Error processing mutation set", e);
            }
```